support nested compile when inner compile is inside of __torch_dispatch__

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/150262

At a high level, the idea is that anywhere we do fake tensor prop inside of compile, there is a risk that this fake prop can recursively invoke compilation. This can happens if we are doing fake prop on a tensor subclass, and it's __torch_dispatch__ uses torch.compile.

Two things to call out:

(1) I took @williamwen42 's advice and used `set_stance("force_eager")`, to ensure that dynamo doesn't turn back on

(2) I needed to use it both in dynamo's fake prop, and in AOTAutograd compilation. I also needed to set `error_on_nested_fx_trace`, since otherwise dynamo will raise an error when it sees it is running from an `fx_trace` context. We could theoretically avoid this if we properly **turned off** dynamo, but `set_stance()` still leaves dynamo on.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150355
* #150302



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bdhirsh committed:
support nested compile when inner compile is inside of __torch_dispatch__

[ghstack-poisoned]

williamwen42 reviewed (approved):
We could bypass the checks done in dynamo eval_frame.py if we determine that we're in force_eager stance. I can do this in a followup.

zou3519 commented on test/dynamo/test_repros.py:
```diff
@@ -7057,6 +7057,52 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+                if kwargs is None:
+                    kwargs = {}
+                if func is torch.ops.aten.add.Tensor:
+                    return torch.compile(func, backend="aot_eager")(
+                        args[0].elem, args[1].elem
+                    )
```
it's not clear to me what to do when the options for the inner compile are different from the outer compile

bdhirsh replied:
it's not clear to me either. Some thoughts though:

(1) we probably want to be consistent with how we normally treat nested compile (in the not-torch_dispatch case). What does dynamo see when it encounters a compiled frame? I imagine it just ignores it and inlines the function?

(2) one thing it feels to me like we have to make a decision on: either the outer compile should inline the inner compile, **or** outer compile should treat the inner compile as a custom op?

One data-point is that in @gau-nernst's example, it sounds like he didn't actually want different compile settings. He had a small quantization subroutine that he wanted **always**-compiled. Then he wanted the ability to optionally wrap the whole model in compile (it had more problems). The inlining in this case seems like a better outcome. I'm not sure if users want the ability to choose between these options though.

(3) there is probably a related composability question here about making compile configs regional

zou3519 reviewed (approved):
seems better than the previous state but idk about composability

bdhirsh commented on torch/_dynamo/utils.py:
```diff
@@ -3137,7 +3137,11 @@ def get_fake_value(node, tx, allow_non_graph_fake=False):
+            tx.fake_mode,
+            enable_python_dispatcher(),
+            torch.compiler.set_stance("force_eager"),
+        ):
```
I guess this error will start happening in any tests that have dynamo run directly on the AOTDispatcher compile-time code. My question is: are we ok with these new failures? Or should we figure out a way for `set_stance()` to not error in this case?

williamwen42 replied:
Is it expected that dynamo should try to trace into AOTDispatcher code here? And should `set_stance` even be applied in this case?

Anyway, we can use `torch.compiler.is_compiling` to only only `set_stance()` when we're not compiling.

zou3519 replied:
we're not OK with these new failures, opcheck needs to be able to run over code that calls torch.compile

bdhirsh replied:
I tried branching on `torch.compiler.is_compiling()`, and that wasn't enough.

I tried branching on `torch._C._dynamo.eval_frame.get_eval_frame_callback() is not None` though (emulating what the error check is doing, and that seemed to work. I guess this implies that `torch.compiler.is_compiling()` returning True is not one-to-one with us setting the dynamo eval frame callback. Does that seem expected?

(I just updated the PR to do this check, although it feels a bit bad, but open to suggestions)

bdhirsh replied:
The other thing I noticed is that just adding the `set_stance("force_eager")` change in `dynamo/utils.py` is enough to regress the small compile time benchmarks by 2% (there are some failures in this PR). I also confirmed that locally.

To be fair, right now this PR calls `set_stance("force_eager")` every time dynamo runs fake prop for every node in the graph.

I'm wondering if you think:

(1) set_stance has some inherent slowness (that either we can fix, or I should try another API)

(2) we can alternative move my call to `set_stance` to happen somewhere higher-level in dynamo, and not per-node getting evaluated

williamwen42 replied:
Ah, `is_compiling` is not the same as checking `get_eval_frame_callback` - for example, if we skip a frame, the callback will still be set, but `is_compiling` would return False. I think it would be fine to have a new function in eval_frame.py to check if the callback is False/None (`eval_frame_callback_enabled`?)

williamwen42 replied:
As for the compile time regression, setting the eval frame callback directly won't be enough since any future compile directive will reset the callback. I would try finding a place higher up in dynamo to place it so that we're not setting stance constantly.

bdhirsh committed:
Update on "support nested compile when inner compile is inside of __torch_dispatch__"

Fixes https://github.com/pytorch/pytorch/issues/150262

At a high level, the idea is that anywhere we do fake tensor prop inside of compile, there is a risk that this fake prop can recursively invoke compilation. This can happens if we are doing fake prop on a tensor subclass, and it's __torch_dispatch__ uses torch.compile.

Two things to call out:

(1) I took williamwen42 's advice and used `set_stance("force_eager")`, to ensure that dynamo doesn't turn back on

(2) I needed to use it both in dynamo's fake prop, and in AOTAutograd compilation. I also needed to set `error_on_nested_fx_trace`, since otherwise dynamo will raise an error when it sees it is running from an `fx_trace` context. We could theoretically avoid this if we properly **turned off** dynamo, but `set_stance()` still leaves dynamo on.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "support nested compile when inner compile is inside of __torch_dispatch__"

Fixes https://github.com/pytorch/pytorch/issues/150262

At a high level, the idea is that anywhere we do fake tensor prop inside of compile, there is a risk that this fake prop can recursively invoke compilation. This can happens if we are doing fake prop on a tensor subclass, and it's __torch_dispatch__ uses torch.compile.

Two things to call out:

(1) I took williamwen42 's advice and used `set_stance("force_eager")`, to ensure that dynamo doesn't turn back on

(2) I needed to use it both in dynamo's fake prop, and in AOTAutograd compilation. I also needed to set `error_on_nested_fx_trace`, since otherwise dynamo will raise an error when it sees it is running from an `fx_trace` context. We could theoretically avoid this if we properly **turned off** dynamo, but `set_stance()` still leaves dynamo on.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]
